code/game/objects/effects/portals.dm

```diff
@@ -45,20 +46,23 @@
 	return ..()
 
 /obj/effect/portal/attackby(obj/item/W, mob/user, params)
-	if(user && Adjacent(user))
-		user.forceMove(get_turf(src))
-		return TRUE
+	if(user)
+		if(get_turf(user) == get_turf(src))
+			teleport(user)
+			return TRUE
+		if(Adjacent(user))
+			user.Move(get_turf(src))
+			return TRUE
 
 /obj/effect/portal/attack_tk(mob/user)
 	return
 
-/obj/effect/portal/proc/on_entered(datum/source, atom/movable/arrived, direction)
-	SIGNAL_HANDLER
-	if(isobserver(arrived))
-		return
-	if(linked && (get_step(source, REVERSE_DIR(direction)) == get_turf(linked)))
```

Comment:
the problem is actually in this line, portals will work fine if and only if they are adjacent, if theyre further than one tile away from each other then get_step() will never be equal to the turf of the linked portal so it will go forever

Reply:
The problem is the old args passed the teleporte's `oldloc` which could be compared to the linked portals loc in order to stop loops.

Reply:
i know, another solution needs to be found now.